### PR TITLE
docs(idempotency): add cdk table example

### LINKED ISSUE #2433: Docs: Idempotency CDK example
### What were you searching in the docs?

I was looking for a CDK idempotency table example but found only the SAM template.

### Is this related to an existing documentation section?

https://awslabs.github.io/aws-lambda-powertools-python/2.16.2/utilities/idempotency/#required-resources

### How can we improve?

Add CDK construct example

### Got a suggestion in mind?

I'll do it :)

### Acknowledgment

- [X] I understand the final update might be different from my proposed suggestion, or refused.

### PR DESCRIPTION
**Pull request by ran-isenberg**
**Issue number: #2433**

## Summary
Added CDK construct example for idempotency setup

### Changes

> Please provide a summary of what's being changed

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [ ] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [ ] I have performed a self-review of this change
* [ ] Changes have been tested
* [ ] Changes are documented
* [ ] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by ran-isenberg** [comment]
please note that the other long lines changes that i did were because the linter just failed. feel free to revet them.
This is the error i got:
- hook id: markdownlint
- exit code: 1

docs/utilities/idempotency.md:361:266 MD013/line-length Line length [Expected: 265; Actual: 273]
docs/utilities/idempotency.md:437:266 MD013/line-length Line length [Expected: 265; Actual: 320]
docs/utilities/idempotency.md:445:266 MD013/line-length Line length [Expected: 265; Actual: 284]

**Comment by rubenfonseca** [comment]
Looking at this now.

File: docs/utilities/idempotency.md [hunk 1]
```diff
@@ -148,7 +185,7 @@ You can quickly start by initializing the `DynamoDBPersistenceLayer` class and u
 
 === "Example event"
 
-    ```json
+   ```json
```
Comment (rubenfonseca) [inline docs/utilities/idempotency.md:136]: I think we removed one space here, can we bring it back so we have 4 spaces indentation?
```suggestion
    ```json
```

File: docs/utilities/idempotency.md [hunk 2]
```diff
@@ -77,31 +77,68 @@ If you're not [changing the default configuration for the DynamoDB persistence l
...
+                'idempotency-policy',
+                statements=[
+                    iam.PolicyStatement(
+                        actions=['dynamodb:PutItem', 'dynamodb:GetItem', 'dynamodb:UpdateItem', 'dynamodb:DeleteItem'],
+                        resources=[self.idempotency_table.table_arn],
+                        effect=iam.Effect.ALLOW,
+                    )
+                ]
+            ))
+    ```
```
Comment (rubenfonseca) [inline docs/utilities/idempotency.md:89]: Can we move the SAM and CDK samples into `examples/` folder?

Context: we are taking every opportunity we get to move code snippets into examples/, where our linting and type checker can run, so we make sure the code is good to be copy/pasted.

File: docs/utilities/idempotency.md [hunk 3]
```diff
@@ -77,31 +77,68 @@ If you're not [changing the default configuration for the DynamoDB persistence l
...
+                self,
+                'idempotency-policy',
+                statements=[
+                    iam.PolicyStatement(
+                        actions=['dynamodb:PutItem', 'dynamodb:GetItem', 'dynamodb:UpdateItem', 'dynamodb:DeleteItem'],
+                        resources=[self.idempotency_table.table_arn],
+                        effect=iam.Effect.ALLOW,
+                    )
+                ]
+            ))
```
Comment (rubenfonseca) [inline docs/utilities/idempotency.md]: Maybe we could use the `grant` functionality of CDK instead?
```suggestion
            self.idempotency_table.grant_read_write_data(lambda_role)
```

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
Fantastic work Ran! Left a couple of comments. Do you mind moving the snippets to `example/`? Let me know if you need a hand with this!

File: docs/utilities/idempotency.md (hunk 3 above)
Comment (ran-isenberg) [reply docs/utilities/idempotency.md]: i dotn usually recommend using the grant_read_write_data as it provides more permissions than you really need (I checked). i wanted it to be least privilege.

Comment (rubenfonseca) [reply docs/utilities/idempotency.md]: I understand. Two asks then:

**1/** Use [`table.grant`](https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_dynamodb/Table.html#aws_cdk.aws_dynamodb.Table.grant) with the permissions you need, as it's more CDK friendly:

```suggestion
            self.idempotency_table.grant(lambda_role, ['dynamodb:PutItem', 'dynamodb:GetItem', 'dynamodb:UpdateItem', 'dynamodb:DeleteItem'])
```

**2/** We will need to update the SAM example for consistency, as it currently uses the overly-permissive `DynamoDBCrudPolicy`.

Let me know if you need a hand for this!

**Comment by ran-isenberg** [comment]
@rubenfonseca  i moved them as snippets. However, sam had plenty of cloudformation lint errors (even though it's not cloudformation). so i pasted an empty file.
i dunno how to fill the sam part anyways.

BTW, i saw that you guys added a vpc lattice model but no parser variation :(

File: docs/utilities/idempotency.md (hunk 2 above)
Comment (ran-isenberg) [reply docs/utilities/idempotency.md:89]: done

File: docs/utilities/idempotency.md (hunk 1 above)
Comment (ran-isenberg) [reply docs/utilities/idempotency.md:136]: i used a regular tab, is that not ok?

Comment (rubenfonseca) [reply docs/utilities/idempotency.md:136]: You're right, please disregard!

**Review by rubenfonseca — APPROVED** [review]

**Comment by rubenfonseca** [comment]
Thank you for this great addition Ran!

**Pull request merged**